### ISSUE TITLE
container hangs after "docker logs <container_name>" is killed

### ISSUE THREAD

**Issue by stszap** [docker/docker#30135]
**Description**

Hello. We switched from 1.12 to 1.13.0_rc6 and after few hours some containers got stuck (could not be stopped, "docker logs" and "docker exec" hanged also). We are constantly deleting and creating containers and also using "docker logs" and "docker exec" a lot.  We are able to reliably reproduce this now.

**Steps to reproduce the issue:**
1. start container, that generate a lot of logs `docker run -d --name test debian cat /dev/urandom`
2. open logs `docker logs test`
3. kill "docker logs" with Ctrl+C
4. stop container `docker stop test` 
5. stop container again `docker stop test` 

I also can reproduce this with simple bash script
```
#!/bin/sh

echo "starting container"
docker run -d --name test debian cat /dev/urandom
sleep 10

echo "requesting logs"
docker logs test > /dev/null &
sleep 10

echo "killing logs"
kill %1
sleep 10

echo "stopping container first time"
docker stop test

echo "stopping container second time"
docker stop test
```

**Describe the results you received:**
Second "docker stop" hangs container for good. "docker exec test", "docker logs test" and "docker inspect test" hangs and "docker ps" also hangs. New containers could not be started.

**Describe the results you expected:**
No hanging.

**Additional information you deem important (e.g. issue happens only occasionally):**
I tested this on 1.12.6 and everything is ok. I also tried docker-machine (virtualbox driver) with boot2docker 1.13.0_rc6 but results were different. Second "docker stop" completes, but "docker ps" shows that container is running. 

**Output of `docker version`:**

```
Client:
 Version:      1.13.0-rc6
 API version:  1.25
 Go version:   go1.7.4
 Git commit:   2f2d055
 Built:        Fri Jan 13 19:48:58 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-rc6
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.4
 Git commit:   2f2d055
 Built:        Fri Jan 13 19:48:58 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**
```
Containers: 34
 Running: 0
 Paused: 0
 Stopped: 34
Images: 490
Server Version: 1.13.0-rc6
Storage Driver: devicemapper
 Pool Name: docker-8:4-5509320-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 10.45 GB
 Data Space Total: 107.4 GB
 Data Space Available: 62.68 GB
 Metadata Space Used: 21.07 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.126 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93 (2015-01-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: v0.2.5 (expected: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e)
runc version: c91b5be (expected: 2f7393a47307a16f8cee44a37b262e8b81021e3e)
init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.12.52-gentoo-r1
Operating System: Gentoo/Linux
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.614 GiB
Name: nio
ID: 6XIR:QBFD:VPGE:A6QA:6EIA:OOE4:CWZB:4ODI:IOZF:THCT:HOP4:AYXS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical
`Linux nio 3.12.52-gentoo-r1 #1 SMP Tue Oct 25 19:41:02 YEKT 2016 x86_64 Intel(R) Pentium(R) CPU G3430 @ 3.30GHz GenuineIntel GNU/Linux`

**Comment by cpuguy83** [comment]
ping @mlaventure

**Comment by mlaventure** [comment]
Hum couldn't reproduce on my system after a few dozen runs. This may be related to changes in devicemapper if we made any.

**Comment by mlaventure** [comment]
Managed to reproduce it on a Fedora box with overlay2, will have a look.

**Comment by Kukunin** [comment]
still have it on `1.13.1`

```
root@w2-reasedoper:~# docker info
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 4
Server Version: 1.13.1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Kernel Version: 4.5.7-std-3
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.956 GiB
Name: w2-reasedoper
ID: 7E5T:F46T:JG7F:EAQS:DTUM:XRC3:5Z5F:CFPV:3DAT:S36F:LBJU:NCYI
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by mlaventure** [comment]
@Kukunin it didn't make it into 1.13.1 unfortunately

**Comment by thiagotrennepohl** [comment]
``` dockerd[7987]: time="2017-02-21T11:36:53.260467066-03:00" level=error msg="containerd: deleting container" error="exit status 1: \"container xxxhxhxhxysysyxhxhxhxhysysyxhxh2838123 does not exist\\none or more of the container deletions failed\\n\""
dockerd[7987]: time="2017-02-21T11:37:03.263605371-03:00" level=warning msg="Logger didn't exit in time: logs may be truncated" ```

**Comment by thiagotrennepohl** [comment]
Same issue here.

```dockerd[7987]: time="2017-02-21T11:36:53.260467066-03:00" level=error msg="containerd: deleting container" error="exit status 1: \"container xxxxxxxxxxxxxxxxxxxxxx does not exist\\none or more of the container deletions failed\\n\""
 dockerd[7987]: time="2017-02-21T11:37:03.263605371-03:00" level=warning msg="Logger didn't exit in time: logs may be truncated"
 dockerd[7987]: time="2017-02-21T11:43:46.401913645-03:00" level=error msg="containerd: deleting container" error="exit status 1: \"container xxxxxxxxxxxxxxxxxxxxxx does not exist\\none or more of the container deletions failed\\n\"```


```Containers: 13
 Running: 13
 Paused: 0
 Stopped: 0
Images: 7
Server Version: 1.13.0
Storage Driver: aufs
 Root Dir: /data/docker/aufs
 Backing Filesystem: extfs
 Dirs: 85
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-53-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.42 GiB
Name: zzzz-zzzz-zzz-zzz
ID: KKLN:YTCT:HWOY:PREL:E7PW:CU2E:T72C:7HUM:NMWX:L5HC:CC4W:7GY7
Docker Root Dir: /data/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by cpuguy83** [comment]
@thiagotrennepohl Thanks for reporting, but this is a closed issue that is resolved and will be fixed in the next patch release.

**Comment by Kukunin** [comment]
@cpuguy83 should we expect the fix in 1.13.2 or so?

**Comment by rmoriz** [comment]
Same here on Docker version 1.13.0, build 49bf474 / Debian / 4.8.0-0.bpo.2-amd64

- docker logs --tail froze/no output
- container hangs
- container can't be stopped
- docker ps still shows "running" state
- contianer can't be killed

**Comment by thaJeztah** [comment]
@rmoriz this issue was fixed and the fix is included in docker 17.03, so you'll have to upgrade to that version to get the fix. If you're still able to reproduce on 17.03, please open a new issue with details.

I'm locking the conversation on this issue, because the original issue reported here was resolved; if you still run into this issue on docker 17.03.0 or above, opening a new issue is the way to report this, otherwise your bug report may go unnoticed.
